README.md
# python-vast-xml-generator
if you plan to write an video ad server with vast support you can generate xml response using this lib, 

## Example
#### Create instance
```python
vast = VAST({"version":"3.0", "VASTErrorURI": "optional url if something went wrong in client side"}) # create instance of vast object, version is vast version, 
```


#### Adding Ad
```python
ad = vast.attachAd({ 
    "id": "1", # ad id 
    "structure": 'inline', # or "wrapper", 
    "sequence": "1", # optional, not required
    "Error": 'http://error.err', # error url if something went wrong in client side, optional
    "AdTitle": 'Common name of the ad' # required for inline structure, 
    "AdSystem": { "name": 'name of adserver or company', "version": "1.0"  }, 
    "Description": "optional description of ad",
    "Advertiser": "Optional name of advertiser",
    "Pricing": "Optional price (if you want to RTB on vast)",
    "Extensions": """<xml>xml extension for client side</xml>""",
  })
```


#### Add Impression url 
```python 
# you can add many servers if you need to
ad.attachImpression({
    "id": "servername"
   "url": "impression url"
})
```
#### Add Linear Creative 
```python
# you can give any valid VAST XmlTagName and value for Creative
creative = ad.attachCreative('Linear', {
    "AdParameters" : """<xml></xml>""", #Optional
    "Duration" : '00:00:30' # required for linear type
});
# you can give any valid VAST XmlTagName and value for media file
creative.attachMediaFile('file url', {
   "type": "video/mp4",
   "bitrate": "320",
   "minBitrate": "320",
   "maxBitrate": "320",
   "width": "640",
   "height": "360",
   "scalable": "true",
   "maintainAspectRatio": "true",
   "codec": "",
   "apiFramework": "VPAID",
});

# You can add any valid  tracking events: vast/trackingEvent.py#L18
creative.attachTrackingEvent('creativeView', 'server url');
# many times 
creative.attachTrackingEvent('progress', 'server url', '00:00:01');
#
creative.attachVideoClick('ClickThrough', 'click target url');
creative.attachClickThrough("Url of server")
creative.attachClick("Url") # look at the vast 3.0 documentation 

vast.xml() // returns XMLBuilder object, to print str(vast.xml()) 

```


CHANGELOG
##  2019-05-31
### Changed
- LICENCE file changed from GPLV3 to APACHE v2 and fixed licence conflicts 
between licence file and headers of .py files
- merged pr#2 this pr will fixed CDATA wrapping and ClickTracking for NonLinear ads.
 
### Added
- added CHANGELOG file




vast/creative.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2015 Timu Eren <timu.eren@gmail.com>
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use self file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from icon import Icon
from trackingEvent import TrackingEvent

VALID_VIDEO_CLICKS = ['ClickThrough', 'ClickTracking', 'CustomClick']


class Creative(object):
    def __init__(self, _type, settings=None):
        settings = {} if settings is None else settings
        self.type = _type
        self.mediaFiles = []
        self.trackingEvents = []
        self.videoClicks = []
        self.clickThroughs = []
        self.clicks = []
        self.resources = []
        self.icons = []
        self.AdParameters = settings.get("AdParameters", None)
        self._adParameters = None
        self.attributes = {}
        self.duration = settings.get("Duration", None)
        self.skipoffset = settings.get("skipoffset", None)
        self.nonLinearClickThrough = None
        self.nonLinearClickTracking = None

        if _type == "Linear" and self.duration is None:
            raise Exception('A Duration is required for all creatives. Consider defaulting to "00:00:00"')

        if "id" in settings:
            self.attributes["id"] = settings["id"]

        if "width" in settings:
            self.attributes["width"] = settings["width"]
        if "height" in settings:
            self.attributes["height"] = settings["height"]
        if "expandedWidth" in settings:
            self.attributes["expandedWidth"] = settings["expandedWidth"]
        if "expandedHeight" in settings:
            self.attributes["expandedHeight"] = settings["expandedHeight"]
        if "scalable" in settings:
            self.attributes["scalable"] = settings["scalable"]
        if "maintainAspectRatio" in settings:
            self.attributes["maintainAspectRatio"] = settings["maintainAspectRatio"]
        if "minSuggestedDuration" in settings:
            self.attributes["minSuggestedDuration"] = settings["minSuggestedDuration"]
        if "apiFramework" in settings:
            self.attributes["apiFramework"] = settings["apiFramework"]

    def attachMediaFile(self, url, settings={}):
        media_file = {"attributes": {}}
        media_file["url"] = url
        media_file["attributes"]["type"] = settings.get("type", 'video/mp4')
        media_file["attributes"]["width"] = settings.get("width", '640')
        media_file["attributes"]["height"] = settings.get("height", '360')
        media_file["attributes"]["delivery"] = settings.get("delivery", 'progressive')
        if "id" not in settings:
            raise Exception('an `id` is required for all media files')

        media_file["attributes"]["id"] = settings["id"]
        if "bitrate" in settings:
            media_file["attributes"]["bitrate"] = settings["bitrate"]
        if "minBitrate" in settings:
            media_file["attributes"]["minBitrate"] = settings["minBitrate"]
        if "maxBitrate" in settings:
            media_file["attributes"]["maxBitrate"] = settings["maxBitrate"]
        if "scalable" in settings:
            media_file["attributes"]["scalable"] = settings["scalable"]
        if "codec" in settings:
            media_file["attributes"]["codec"] = settings["codec"]
        if "apiFramework" in settings:
            media_file["attributes"]["apiFramework"] = settings["apiFramework"]
        if "maintainAspectRatio" in settings:
            media_file["attributes"]["maintainAspectRatio"] = settings["maintainAspectRatio"]

        self.mediaFiles.append(media_file)
        return self

    def attachTrackingEvent(self, _type, url, offset=None):
        self.trackingEvents.append(TrackingEvent(_type, url, offset))
        return self

    def attachVideoClick(self, _type, url, _id=''):
        if _type not in VALID_VIDEO_CLICKS:
            raise Exception('The supplied VideoClick `type` is not a valid VAST VideoClick type.')
        self.videoClicks.append({"type": _type, "url": url, "id": _id})
        return self

    def attachClickThrough(self, url):
        self.clickThroughs.append(url)
        return self

    def attachClick(self, uri, _type=None):
        if isinstance(uri, basestring):
            _type = 'NonLinearClickThrough'
        self.clicks = [{"type": _type, "uri": uri}]
        return self

    def attachResource(self, _type, uri, creative_type=None):
        resource = {"type": _type, "uri": uri}
        if _type == 'HTMLResource':
            resource["html"] = uri
        if creative_type is not None:
            resource["creativeType"] = creative_type
        self.resources.append(resource)
        return self

    def attachIcon(self, settings):
        icon = Icon(settings)
        self.icons.append(icon)
        return icon

    def adParameters(self, data, xml_encoded):
        self._adParameters = {"data": data, "xmlEncoded": xml_encoded}
        return self

    def attachNonLinearClickThrough(self, url):
        self.nonLinearClickThrough = url

    def attachNonLinearClickTracking(self, url):
        self.nonLinearClickTracking = url
